### PR TITLE
added mechanism to break if not enough time remained for another fetc…

### PR DESCRIPTION
**Pull request by YuvHayun**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [XSUP-43176](https://jira-dc.paloaltonetworks.com/browse/XSUP-43176)

### CODE REVIEW

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 1]
```diff
@@ -35,7 +35,11 @@
 
 VENDOR = "Akamai"
 PRODUCT = "WAF"
-FETCH_EVENTS_PAGE_SIZE = 50000
+FETCH_EVENTS_PAGE_SIZE = 15000
+DOCKER_MIN_TIME_TO_RUN = 30
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:39]:
```suggestion
TIME_TO_RUN_BUFFER = 30
```

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 2]
```diff
@@ -35,7 +35,11 @@
 
 VENDOR = "Akamai"
 PRODUCT = "WAF"
-FETCH_EVENTS_PAGE_SIZE = 50000
+FETCH_EVENTS_PAGE_SIZE = 15000
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:38]:
```suggestion
FETCH_EVENTS_MAX_PAGE_SIZE = 15000
```

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 3]
```diff
@@ -35,7 +35,11 @@
...
+DOCKER_MIN_TIME_TO_RUN = 30
+EXECUTION_START_TIME = datetime.now()
+ALLOWED_PAGE_SIZE_DELTA_RATIO = 0.95
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:41]: add a comment with an explanation

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 4]
```diff
@@ -405,6 +409,44 @@ def dedup_events(hashed_events_mapping: dict[str, dict], hashed_events_from_prev
...
+    The remaining running time is docker timeout limit in seconds - the run time so far (now time - docker execution start time).
+
+    Args:
+        min_allowed_delta (int): The minimum allowed delta that should remain before going on another fetch interval.
+        max_time_took (float): The worst case execution (the first execution) to compare the rest of the executions to.
+    Returns:
+        bool: Return True if there's not enough time. Otherwise, return False.
+    """
+    timeout_time_nano_seconds = demisto.callingContext.get('context', {}).get('TimeoutDuration')
+    timeout_time_seconds = timeout_time_nano_seconds / 1000000000
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:441]:
```suggestion
    timeout_time_seconds = timeout_time_nano_seconds / 1_000_000_000
```

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 5]
```diff
@@ -405,6 +409,44 @@ def dedup_events(hashed_events_mapping: dict[str, dict], hashed_events_from_prev
...
+    And checks wether the remaining running time (plus a little delta) is less or equal the expected running time.
+    The remaining running time is docker timeout limit in seconds - the run time so far (now time - docker execution start time).
+
+    Args:
+        min_allowed_delta (int): The minimum allowed delta that should remain before going on another fetch interval.
+        max_time_took (float): The worst case execution (the first execution) to compare the rest of the executions to.
+    Returns:
+        bool: Return True if there's not enough time. Otherwise, return False.
+    """
+    timeout_time_nano_seconds = demisto.callingContext.get('context', {}).get('TimeoutDuration')
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:440]: log this value

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 6]
```diff
@@ -425,15 +467,43 @@ def fetch_events_command(
...
+                break
+        if (remaining_events_to_fetch := fetch_limit - total_events_count) < page_size:
+            demisto.info(f"{remaining_events_to_fetch=} < {page_size=}, lowering page_size to {remaining_events_to_fetch}.")
+            page_size = remaining_events_to_fetch
         demisto.info(f"Preparing to get events with {offset=}, {page_size=}, and {fetch_limit=}")
-        events, offset = client.get_events_with_offset(config_ids, offset, page_size, from_epoch)
+        try:
+            events, offset = client.get_events_with_offset(config_ids, offset, page_size, from_epoch)
+        except DemistoException as e:
+            if "Requested Range Not Satisfiable" in str(e):
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:498]: log the fact that there was an error in all cases

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py [hunk 7]
```diff
@@ -521,23 +592,30 @@ def main():  # pragma: no cover
...
                 if events:
                     demisto.info(f"Sending events to xsiam with latest event time is: {events[-1]['_time']}")
                     send_events_to_xsiam(events, VENDOR, PRODUCT, should_update_health_module=False)
                 set_integration_context({"offset": offset,
                                          "hashed_events_from_previous_run": list(hashed_events_from_current_run)})
             demisto.updateModuleHealth({'eventsPulled': (total_events_count or 0)})
             next_run = {}
-            if total_events_count >= limit:
-                demisto.info(f"got at least {limit} events this interval - will automatically trigger next run.")
+            if auto_trigger_next_run or total_events_count >= limit:
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:617]: shouldnt we check if its within the ratio? Meaning getting 44999 events should still trigger next run

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.yml [hunk 8]
```diff
@@ -72,15 +72,15 @@ configuration:
...
+  additionalinfo: Limit on the number of incidents retrieved in a single fetch. The maximum is 45k.
   section: Collect
 - display: Akamai Page size
   name: page_size
   type: 0
   required: false
   section: Collect
-  defaultvalue: '50000'
-  additionalinfo: The number of events to fetch per request to akamai (multiple requests are made for each fetch). If you're getting aggregated delays, increase the number. The maximum is 600,000.
+  defaultvalue: '15000'
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.yml:82]: Add a note about how the integration will catch up in times of bulk. Run this line by me before final merge

File: Packs/Akamai_SIEM/ReleaseNotes/1_1_9.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+#### Integrations
+
+##### Akamai WAF SIEM
+***breaking Changes:*** Added limitations to **Fetch limit** and **page size** parameters due to Restrictions from the platform. The maximum allowed amount is 45k. Note that this change will not fail the execution, but if higher rates are configured, the fetch will use 45k.
```
Comment (JudahSchwartz) [inline Packs/Akamai_SIEM/ReleaseNotes/1_1_9.md:4]: i dont think we need to release it as breaking change as it should only improve performance. Add a jnote about how if the maximum is brought the next run will start immediately (add this to the docs above too)

**Review by JudahSchwartz — APPROVED** [review]
nice!

File: Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py (hunk 7 above)
Comment (YuvHayun) [reply Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py:617]: If the request returned a page lower by delta then we don't want to trigger it.
However, if there was a break due to being afraid of timeouts then auto_trigger_next_run should be True.

Anyway, we should be covered for this one.

**Pull request closed without merging**